fvm/transactionInvoker.go

```diff
@@ -296,3 +201,106 @@ func (i TransactionInvoker) normalExecution(
 
 	return
 }
+
+// Clear changes and try to deduct fees again.
+func (i TransactionInvoker) errorExecution(
+	proc *TransactionProcedure,
+	txnState *state.TransactionState,
+	nestedTxnId state.NestedTransactionId,
+	env environment.Environment,
+	errs *errors.ErrorsCollector,
+) error {
+	txnState.DisableAllLimitEnforcements()
+	defer txnState.EnableAllLimitEnforcements()
+
+	// log transaction as failed
+	env.Logger().Info().
+		Err(errs.ErrorOrNil()).
+		Msg("transaction executed with error")
+
+	env.Reset()
+
+	// drop delta since transaction failed
+	restartErr := txnState.RestartNestedTransaction(nestedTxnId)
+	if errs.Collect(restartErr) {
+		return errs.ErrorOrNil()
+	}
+
+	// try to deduct fees again, to get the fee deduction events
+	feesError := i.deductTransactionFees(proc, txnState, env)
+
+	// if fee deduction fails just do clean up and exit
+	if feesError != nil {
+		env.Logger().Info().
+			Err(feesError).
+			Msg("transaction fee deduction executed with error")
+
+		if errs.Collect(feesError) {
+			return errs.ErrorOrNil()
+		}
+
+		// drop delta
+		restartErr = txnState.RestartNestedTransaction(nestedTxnId)
+		if errs.Collect(restartErr) {
```

Comment:
note the minor change here.  we used to ignore this error.  now we collect and return it.  (in reality, this error should never occur)